### PR TITLE
Add Module controller

### PR DESCRIPTION
**Pull request by shahincsejnu**

### CODE REVIEW

File: controllers/util.go [hunk 1]
```diff
@@ -126,123 +115,188 @@ func reconcile(rClient client.Client, ctx context.Context, gv schema.GroupVersio
...
-	source := moduleObj.Spec.ModuleRef.TfMarketplace
-	providerName := moduleObj.Spec.Provider.Name
-	providerSource := moduleObj.Spec.Provider.Source
+	src := source
+
+	sourceSlice := strings.Split(source, "/")
+	if len(sourceSlice) == 0 {
+		return fmt.Errorf("given github repo source link is invalid")
+	}
+	hostName := sourceSlice[0]
```
Comment (faem) [inline controllers/util.go:166]: use url.Parse()

File: controllers/util.go [hunk 2]
```diff
@@ -126,123 +115,188 @@ func reconcile(rClient client.Client, ctx context.Context, gv schema.GroupVersio
...
 	err = ioutil.WriteFile(mainFile, mainTfJson, 0777)
 	if err != nil {
-		fmt.Println("111")
 		return err
 	}
-	err = generateOutputTFFile(outputFile, moduleName, gv, obj)
-	fmt.Println("before terraformInit")
+
+	jsnSchemaPropsForOutput := moduleDefObj.Spec.Schema.Properties["output"]
+	err = generateOutputTFFile(outputFile, moduleName, jsnSchemaPropsForOutput)
```
Comment (faem) [inline controllers/util.go:196]: handle error

File: controllers/util.go [hunk 3]
```diff
@@ -126,123 +115,188 @@ func reconcile(rClient client.Client, ctx context.Context, gv schema.GroupVersio
...
+}
+
+func createGitRepoTempPath(path string) error {
+	_, err := os.Stat(path)
+	if !os.IsNotExist(err) {
+		err = os.RemoveAll(path)
+		if err != nil {
+			return err
+		}
+	}
```
Comment (faem) [inline controllers/util.go:254]: We don't want clone in every reconciliation, if the repo already exists, use that directly.

File: controllers/util.go [hunk 4]
```diff
@@ -320,9 +371,9 @@ func initialUpdateStatus(rClient client.Client, ctx context.Context, gv schema.G
 	typedStruct := structs.New(typedObj)
 	conditionsVal := reflect.ValueOf(typedStruct.Field("Status").Field("Conditions").Value())
 	conditions := conditionsVal.Interface().([]kmapi.Condition)
-	if kmapi.HasCondition(conditions, "Stalled") {
-		return nil
-	}
+	//if kmapi.HasCondition(conditions, "Stalled") {
+	//	return nil
+	//}
```
Comment (faem) [inline controllers/util.go:376]: delete

File: Dockerfile [hunk 5]
```diff
@@ -19,9 +19,20 @@ RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -a -o manager main.go
...
+  && apt-get update \
+  && apt-get install -y --no-install-recommends apt-transport-https ca-certificates curl unzip git openssh-server
+
+RUN set -x \
+  && curl -O -fsSL https://releases.hashicorp.com/terraform/1.1.5/terraform_1.1.5_linux_amd64.zip \
+  && unzip terraform_1.1.5_linux_amd64.zip \
+  && chmod 755 terraform \
+  && mv terraform /bin/terraform
+
+#USER 65532:65532
```
Comment (faem) [inline Dockerfile:36]: delete

File: Dockerfile [hunk 6]
```diff
@@ -19,9 +19,20 @@ RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -a -o manager main.go
 
 # Use distroless as minimal base image to package the manager binary
 # Refer to https://github.com/GoogleContainerTools/distroless for more details
-FROM gcr.io/distroless/static:nonroot
+FROM ubuntu:latest
```
Comment (faem) [inline Dockerfile:22]: ubuntu:20.04

**Review by faem — CHANGES_REQUESTED** [review]

File: controllers/util.go (hunk 3 above)
Comment (shahincsejnu) [reply controllers/util.go]: Done

**Review by faem — APPROVED** [review]

**Pull request merged**